Makefile

```diff
@@ -138,61 +138,71 @@ $(OUT)/test_sifive_u: ${TEST_SIFIVE_U_DEPS}
 $(OUT)/user_sifive_u: ${USER_SIFIVE_U_DEPS}
 	$(RISCV64_GCC) -march=rv64g -mabi=lp64 $(GCC_FLAGS) \
 		-Wa,--defsym,NUM_HARTS=2 \
+		-include include/machine/qemu.h \
 		${USER_SIFIVE_U_DEPS} -o $@
 
 $(OUT)/test_sifive_u32: ${TEST_SIFIVE_U32_DEPS}
 	$(RISCV64_GCC) -march=rv32g -mabi=ilp32 $(GCC_FLAGS) \
 		-Wa,--defsym,XLEN=32 \
 		-Wa,--defsym,NUM_HARTS=1 \
+		-include include/machine/qemu.h \
 		${TEST_SIFIVE_U32_DEPS} -o $@
 
 $(OUT)/user_sifive_u32: ${USER_SIFIVE_U32_DEPS}
 	$(RISCV64_GCC) -march=rv32g -mabi=ilp32 $(GCC_FLAGS) \
 		-Wa,--defsym,XLEN=32 \
 		-Wa,--defsym,NUM_HARTS=2 \
+		-include include/machine/qemu.h \
 		${USER_SIFIVE_U32_DEPS} -o $@
 
 $(OUT)/test_sifive_e: ${TEST_SIFIVE_E_DEPS}
 	$(RISCV64_GCC) -march=rv64g -mabi=lp64 $(GCC_FLAGS) \
 		-Wl,--defsym,ROM_START=0x20400000 -Wa,--defsym,UART=0x10013000 \
 		-Wa,--defsym,NUM_HARTS=1 \
+		-include include/machine/qemu.h \
 		${TEST_SIFIVE_E_DEPS} -o $@
 
 $(OUT)/user_sifive_e: ${USER_SIFIVE_E_DEPS}
 	$(RISCV64_GCC) -march=rv64g -mabi=lp64 $(GCC_FLAGS) \
 		-Wl,--defsym,ROM_START=0x20400000 -Wa,--defsym,UART=0x10013000 \
 		-Wa,--defsym,NUM_HARTS=1 \
+		-include include/machine/qemu.h \
 		${USER_SIFIVE_E_DEPS} -o $@
 
 $(OUT)/test_sifive_e32: ${TEST_SIFIVE_E32_DEPS}
 	$(RISCV64_GCC) -march=rv32g -mabi=ilp32 $(GCC_FLAGS) \
 		-Wl,--defsym,ROM_START=0x20400000 -Wa,--defsym,UART=0x10013000 \
 		-Wa,--defsym,XLEN=32 \
 		-Wa,--defsym,NUM_HARTS=1 \
+		-include include/machine/qemu.h \
 		${TEST_SIFIVE_E32_DEPS} -o $@
 
 $(OUT)/user_sifive_e32: ${USER_SIFIVE_E32_DEPS}
 	$(RISCV64_GCC) -march=rv32g -mabi=ilp32 $(GCC_FLAGS) \
 		-Wl,--defsym,ROM_START=0x20400000 -Wa,--defsym,UART=0x10013000 \
 		-Wa,--defsym,XLEN=32 \
 		-Wa,--defsym,NUM_HARTS=1 \
+		-include include/machine/qemu.h \
 		${USER_SIFIVE_E32_DEPS} -o $@
 
 $(OUT)/test_virt: ${TEST_VIRT_DEPS}
 	$(RISCV64_GCC) -march=rv64g -mabi=lp64 $(GCC_FLAGS) \
 		-Wa,--defsym,UART=0x10000000 -Wa,--defsym,QEMU_EXIT=0x100000 \
+		-include include/machine/qemu.h \
 		${TEST_VIRT_DEPS} -o $@
 
 $(OUT)/user_virt: ${USER_VIRT_DEPS}
 	$(RISCV64_GCC) -march=rv64g -mabi=lp64 $(GCC_FLAGS) \
 		-Wa,--defsym,UART=0x10000000 -Wa,--defsym,QEMU_EXIT=0x100000 \
 		-Wa,--defsym,NUM_HARTS=1 \
+		-include include/machine/qemu.h \
 		${USER_VIRT_DEPS} -o $@
 
 $(OUT)/user_hifive1_revb: ${USER_SIFIVE_E32_DEPS}
 	$(RISCV64_GCC) -march=rv32imac -mabi=ilp32 $(GCC_FLAGS) \
```

Comment:
rv32imac "c" causes "compressed" 2byte per instruction code generation. Should we keep all targets on the same code density for debugability?

Reply:
I'll give it an extra check whether HiFive actually _requires_ it or was it just a default in Freedom SDK. For now, I'm feeling adventurous is going ahead with extra flexibility - this will catch more bugs, but I'm open to idea to unify it back again if it ends up being a nuisance.

Reply:
I gave it a try, and it works well without the compressed instructions as well, i.e. with `-march=rv32ima`. But let's try to keep it compressed, for sportsmanship.